File: day_01/solve.py

def load_file(path: str) -> list[int, int]:
    with open(path, "r") as f:
        lines = f.readlines()
        values = [tuple(int(x) for x in line.split("   ")) for line in lines]
        l1 = [x[0] for x in values]
        l2 = [x[1] for x in values]
        return [l1, l2]

[l1, l2] = load_file("./input.txt")
# print(l1)
# print(l2)

l1.sort()
l2.sort()

distances = [abs(x - y) for x, y in zip(l1, l2)]
# print(distances)

print("Solution to part 1:", sum(distances))

m2 = {} 
for x in l2:
    m2[x] = m2.get(x, 0) + 1

similarities = [x * m2.get(x, 0) for x in l1]
# print(similarities)
print("Solution to part 2:", sum(similarities))
# print(sum)

File: day_01/solve.ts

// Load input file 
async function loadFile(path: string) {
  const text = await Deno.readTextFile(path);
  const data = text
    .split("\n")
    .filter(line => line.trim() !== "")
    .map(line => line.split("   ").map(Number));
  const l1 = data.map(x => x[0]);
  const l2 = data.map(x => x[1]);
  return [l1, l2];
}


const [l1, l2] = await loadFile("./input.txt");
l1.sort();
l2.sort();

// console.log(l1);
// console.log(l2);

const distances = l1.map((x, i) => Math.abs(l2[i] - x));
// console.log(distances);

const sum = distances.reduce((acc, x) => acc + x, 0);
console.log("Solution to part 1:", sum);

const m2 = new Map<number, number>();
for (const x of l2) {
  m2.set(x, (m2.get(x) || 0) + 1);
}
// console.log(m2);

const similarities = l1.map(x => (m2.get(x) || 0) * x);
// console.log(similarities);

const simscore = similarities.reduce((acc, x) => acc + x, 0);
console.log("Solution to part 2:", simscore);
